protocols/imap/src/main/java/org/apache/james/imap/processor/base/UidMsnMapper.java

```diff
@@ -75,7 +79,41 @@ public synchronized void clear() {
     }
 
     public synchronized void addUid(MessageUid uid) {
-        this.addMapping(nextMsn(), uid);
+        if (isLastUid(uid)) {
+            addMapping(nextMsn(), uid);
+        } else {
+            addUidInMiddle(uid);
+        }
+    }
+
+    private boolean isLastUid(MessageUid uid) {
+        return msnToUid.isEmpty() ||
+            uid.asLong() > msnToUid.get(getLastMsn()).asLong();
+    }
+
+    private boolean alreadyContains(MessageUid uid) {
+        return msnToUid.containsValue(uid);
+    }
+
+    private void addUidInMiddle(MessageUid uid) {
+        List<MessageUid> aboveUids = removeAndGetAboveUidSortedInIncreasingOrder(uid);
+        addMapping(nextMsn(), uid);
+        for (MessageUid aboveUid : aboveUids) {
+            addMapping(nextMsn(), aboveUid);
+        }
+    }
+
+    private List<MessageUid> removeAndGetAboveUidSortedInIncreasingOrder(MessageUid uid) {
+        ImmutableList.Builder<MessageUid> result = ImmutableList.builder();
+        int position = getLastMsn();
+        Optional<MessageUid> maxUid = getUid(position);
+        while (maxUid.isPresent() && uid.asLong() < maxUid.get().asLong()) {
```

Comment:
Looking at Guava we can implement this with a simple ArrayList<MessageUid> that we keep sorted

(Collections::sort)

Then we can call Ordering::binarySearches on it.

I just fear we have O(n) insertion or removal time, wich will not be acceptable for 100.000 messages mailboxes. (Note that we can handle the initial construction building in O(n) if we pass the iterator to UidMsnConverter, initial build is not the burden here, but updates upon events are).

The above would perform as is:
 - adduid: O(n) (almost sorted collection)
 - removeUid : O(n) (idem)
 - getUid : O(1)
 - getMsn : O(ln(n)) binarySearch 

Whereas currently implemented algorithm behaves as:
 - addUid() : O(n) but as updates are almost in order we can assume O(1) in most case
 - removeUid() : O(n)
 - getUid : O(1)
 - getMsn: O(1)

I will perform some simple benchs on both implementations to have a more precise idea about it.